File: src/main/java/blog/typed/javadsl/Coexistence1.java
package blog.typed.javadsl;

import akka.actor.AbstractActor;
import akka.typed.ActorRef;
import akka.typed.Behavior;
import akka.typed.javadsl.Adapter;
import static akka.typed.javadsl.Actor.same;


public abstract class Coexistence1 {

  private Coexistence1() {
  }

  public static class MyUntyped1 extends AbstractActor {

    public static akka.actor.Props props() {
      return akka.actor.Props.create(MyUntyped1.class);
    }

    private final akka.typed.ActorRef<MyTyped1.Command> second =
        Adapter.spawn(getContext(), MyTyped1.behavior(), "second");

    @Override
    public void preStart() {
      Adapter.watch(getContext(), second);
      second.tell(new MyTyped1.Ping(Adapter.toTyped(getSelf())));
    }

    @Override
    public Receive createReceive() {
      return receiveBuilder()
        .match(MyTyped1.Pong.class, msg -> {
          System.out.println(getSelf() + " got Pong from " + getSender());
          Adapter.stop(getContext(), second);
        })
        .match(akka.actor.Terminated.class, t -> {
          System.out.println(getSelf() + " observed termination of " + t.getActor());
          getContext().stop(getSelf());
        })
        .build();
    }
  }

  public static abstract class MyTyped1 {
    private MyTyped1(){
    }

    interface Command {
    }

    public static class Ping implements Command {
      public final akka.typed.ActorRef<Pong> replyTo;

      public Ping(ActorRef<Pong> replyTo) {
        this.replyTo = replyTo;
      }
    }

    public static class Pong {
    }

    public static Behavior<Command> behavior() {
      return akka.typed.javadsl.Actor.immutable(MyTyped1.Command.class)
        .onMessage(MyTyped1.Ping.class, (ctx, msg) -> {
          System.out.println(ctx.getSelf() + " got Ping from " + msg.replyTo);
          msg.replyTo.tell(new Pong());
          return same();
        })
        .build();
    }

  }

}

File: src/main/scala/blog/typed/scaladsl/Coexistence1.scala
package blog.typed.scaladsl

import scala.io.StdIn

import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Actor.same

object Coexistence1 {
  import akka.actor.typed.scaladsl.adapter._

  object MyUntyped1 {
    def props(): akka.actor.Props = akka.actor.Props(new MyUntyped1)
  }

  class MyUntyped1 extends akka.actor.Actor {

    // context.spawn is an implicit extension method
    val second: akka.actor.typed.ActorRef[MyTyped1.Command] =
      context.spawn(MyTyped1.behavior, "second")

    // context.watch is an implicit extension method
    context.watch(second)

    // self can be used as the `replyTo` parameter here because
    // there is an implicit conversion from akka.actor.ActorRef to
    // akka.actor.typed.ActorRef
    second ! MyTyped1.Ping(self)

    override def receive = {
      case MyTyped1.Pong =>
        println(s"$self got Pong from ${sender()}")
        // context.stop is an implicit extension method
        context.stop(second)
      case akka.actor.Terminated(ref) =>
        println(s"$self observed termination of $ref")
        context.stop(self)
    }

  }

  object MyTyped1 {
    sealed trait Command
    final case class Ping(replyTo: akka.actor.typed.ActorRef[Pong.type]) extends Command
    case object Pong

    val behavior: Behavior[Command] =
      akka.actor.typed.scaladsl.Actor.immutable { (ctx, msg) =>
        msg match {
          case Ping(replyTo) =>
            println(s"${ctx.self} got Ping from $replyTo")
            replyTo ! Pong
            same
        }
      }
  }

}

object CoexistenceApp1 {
  import Coexistence1._

  def main(args: Array[String]): Unit = {
    val system = akka.actor.ActorSystem("sys")
    system.actorOf(MyUntyped1.props(), "first")
    try {
      // Exit the system after ENTER is pressed
      StdIn.readLine()
    } finally {
      system.terminate()
    }
  }
}
